File: baekjoon/src/tags/queue/No2164.java
package tags.queue;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Deque;
import java.util.LinkedList;

/**
 * https://www.acmicpc.net/problem/2164
 */
public class No2164 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        Deque<Integer> deque = new LinkedList<>();
        int input = Integer.parseInt(reader.readLine());
        for (int i = 1; i <= input; i++)
            deque.addLast(i);

        while (deque.size() > 1 && (deque.pollFirst() != 0))
            deque.addLast(deque.pollFirst());

        System.out.println(deque.poll());
    }
}

File: baekjoon/src/tags/data_structures/No2164.kt
package tags.data_structures

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.LinkedList

/**
 * https://www.acmicpc.net/problem/2164
 */
fun main() {
    val reader = BufferedReader(InputStreamReader(System.`in`))
    val writer = BufferedWriter(OutputStreamWriter(System.out))
    val queue = LinkedList<Int>()

    val numberOfInput = reader.readLine().toInt()
    for (number in 1..numberOfInput)
        queue.add(number)

    while (queue.size != 1) {
        queue.pop()
        queue.add(queue.pop())
    }

    writer.write("${queue.pop()}")
    writer.flush()
}
